programs/wineboot/wineboot.c

```diff
@@ -245,7 +245,7 @@ static void initialize_xstate_features(struct _KUSER_SHARED_DATA *data)
 
 static UINT64 read_tsc_frequency(void)
 {
-    UINT64 freq = 0;
+    static UINT64 freq = 0;
```

Comment:
The helper was initially used only once because the frequency was set in USD page.

As I've added the HV page later (to make sure USD and QPC freq is fixed at 10MHz as some code may expect it), it's been called twice as the HV page may benefit from a better precision.

IMHO it would be better to compute the frequency once keeping as much precision as possible and use it to set the various locations where it's used, instead of having a static variable.

Reply:
Also it's only going to work with Proton Experimental, there's no TSC frequency computation in Proton 6.3 yet (and I'm not sure there's going to soon, as I'm not completely satisfied with the calibration). I'm looking for ways to get it from the kernel, but there's no really good way to do so.

Reply:
What if this could be made overwritable via an environment variable? People that feel like overriding the auto-detected value with one specific to their hardware could do so then.

Reply:
Well, we can easily bring that function over for 6.3.

But yeah, it's sad there's no way of getting it, there were some patches for it but nothing ever in mainline.
Maybe we could have something for it in our own and community kernels though.

Reply:
Made it calculate once at the start of wineboot and pass it to the right places instead of static-ing it

Reply:
@kakra Might be worth doing as a separate patch, this TSC stuff is already used elsewhere.